0000/README.adoc

```diff
@@ -0,0 +1,175 @@
+how a Java object can be mapped to a database, while https://jcp.org/en/jsr/detail?id=222[JAXB] define
+annotations to customize how java attribute will be mapped to xml elements.
+
+This JEP is about introducing support for annotation-based declarative data binding. A DataBound jenkins
+component would only rely on annotation to define how attributes are bound to JSON payload from web forms,
+including validation constraints. We consider this would have two major benefits :
+
+1. Plugin developers will need to write less code, and use standard annotations they probably already know
+from other Java related developments.
+
+2. Constraints on the data model can be introspected. This will allow third party tools to adjust their
+behavior accordingly. A good candidates for this is Configuration-as-Code (JEP-201). Jelly tags use to
+compose the web UI also could introspect those annotations. Typically, if some attribute is a number and
+has to be positive, `<input type='number' 'min=0'>` html 5 element would be automatically generated.
+
+==== Relying on Bean Validation
+
+https://jcp.org/en/jsr/detail?id=222[Bean Validation] is an annotation-based framework designed for
+generic usage defining constraints validation on data in a Java model. It allows definition of model
+constraints in a fully declarative way, and define the mechanism to run validation and discover violations.
+
+Bean Validation also is extensible by developers who can define custom annotations and rules to implement
+them.
+
+==== Relying on JSON-B
+
+http://json-b.net/[Json-B] (JSR 367) is the approved specification for JSON binding. It has been highly
+inspired by https://github.com/FasterXML/jackson[Jackson] and do define the exact binding requirements
+Stapler do implement, but relying on standard annotations (vs custom `@DataBoundConstructor`.
+
+As Stapler will still need to support legacy data-binding for backward compatibility, we can't just replace
+it with a Json-B implementation. We could consider adding support for such standard annotations as
+alternatives to stapler specific ones. But on the other way a Jenkins databound component is nothing like
+a reusable piece of software, and the https://javadoc.io/doc/javax.json.bind/javax.json.bind-api/1.0[Json-B annotations]
+are mostly designed to tune Java to Json conversion, not providing huge benefits.
+
+==== Filed vs Setter injection
+
+An endless debate for annotation based framework is about using annotations on private fields, vs using them
+on accessors. Most frameworks do support both as there's no single answer to this debate.
+
+On one side, injection on private field require Java reflexion to unlock private field accessibility (or
+use Variable Handles on java 9+). It makes testing harder as there's no trivial way to mock or instantiate
+the target component.
+
+On the other side, setter injection require some boilerplate code being added to codebase, and don't prevent
```

Comment:
Annotating `public` fields (as `@DataBoundSetter`, despite its name, supports) avoids both boilerplate _and_ reflection. Of course accessible fields make for poor Java APIs; but as you note in the next paragraph, the classes we are discussing here are not intended as Java APIs.